pkg/kubelet/container/cache.go

```diff
@@ -0,0 +1,187 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package container
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/types"
+)
+
+// Cache stores the PodStatus for the pods. It represents *all* the visible
+// pods/containers in the container runtime. All cache entries are at least as
+// new or newer than the global timestamp (set by SetGlobalTime()), while
+// individual entries may be slightly newer than the global timestamp. If a pod
+// has no states known by the runtime, Cache returns an empty PodStatus object
+// with ID populated.
+//
+// Cache provides two methods to retrive the PodStatus: the non-blocking Get()
+// and the blocking GetNewerThan() method. The component responsible for
+// populating the cache is expected to call Set() to set the status to empty to
+// reclaim the cache entries.
+type Cache interface {
+	Get(types.UID) (*PodStatus, error)
+	Set(types.UID, *PodStatus, error, time.Time)
+	// GetNewerThan is a blocking call that would return the status
+	// when it is newer than the given time.
+	GetNewerThan(types.UID, time.Time) (*PodStatus, error)
+	SetGlobalTimestamp(time.Time)
+}
+
+type data struct {
+	// Status of the pod.
+	status *PodStatus
+	// Error got when trying to inspect the pod.
+	err error
+	// Time when the data was last modfied.
+	modified time.Time
+}
+
+type callbackRecord struct {
+	time time.Time
+	ch   chan *data
+}
+
+// cache implements Cache. It implements the blocking GetNewerThan() call by
+// the callback mechanism.
+type cache struct {
+	// Lock which guards all internal data structures.
+	lock sync.RWMutex
+	// Map that stores the pod statuses.
+	pods map[types.UID]*data
+	// A global timestamp represents how fresh the cached data is. All
+	// cache content is at the least newer than this timestamp. Note that the
+	// timestamp is nil after initialization, and will only become non-nil when
+	// it is ready to serve the cached statuses.
+	timestamp *time.Time
+	// Map that stores the callback records.
+	callbacks map[types.UID]callbackRecord
+}
+
+// NewCache creates a pod cache.
+func NewCache() Cache {
+	return &cache{pods: map[types.UID]*data{}, callbacks: map[types.UID]callbackRecord{}}
+}
+
+// getIfNewerThan returns the data it is newer than the given time.
+// Otherwise, it returns nil.
+func (c *cache) getIfNewerThan(id types.UID, minTime time.Time) *data {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	d, ok := c.pods[id]
+
+	if !ok {
+		// Status is not cached.
+		if c.timestamp != nil && c.timestamp.After(minTime) {
```

Comment:
I still think we can avoid the nil pointer by using time.Time instead of *time.Time in the cache struct.

We can use IsZero() bool to test if the timestamp is set.